File: app/models/stock_trend.rb
class StockTrend < ApplicationRecord
  include Braindamage::Braindamageable
  belongs_to :stock
  expose :stock_prices, type: :has_many, model: 'StockPrice'

  def stock_prices
    stock.stock_prices.where('day >= ?', started_at)
  end
end

# == Schema Information
#
# Table name: stock_trends
#
#  id          :bigint           not null, primary key
#  deviation   :decimal(15, 8)
#  finished_at :date
#  intercept   :decimal(15, 8)
#  slope       :decimal(15, 8)
#  started_at  :date             not null
#  created_at  :datetime         not null
#  updated_at  :datetime         not null
#  stock_id    :bigint           not null
#
# Indexes
#
#  index_stock_trends_on_stock_id  (stock_id)
#
# Foreign Keys
#
#  fk_rails_...  (stock_id => stocks.id)
#

File: app/javascript/models/stock_trend.js
import Model from "braindamage/model";
import StockTrendSchema from 'generated/schemas/stock_trend';
import { getTime } from "date-fns";

class StockTrend extends Model {
  static schema = StockTrendSchema;
  static modelName = 'StockTrend';

  // Fill-in your attribute overrides
  static attributesDefinitions() {
    return {};
  }

  dayToX(day) {
    return getTime(day) / (1000 * 60 * 60 * 24);
  }

  y(x) {
    return this.slope * x + this.intercept;
  }

  priceAt(day = new Date()) {
    return this.y(this.dayToX(day));
  }

  supportPrice(day = new Date()) {
    return this.priceAt(day) - this.deviation;
  }

  resistancePrice(day = new Date()) {
    return this.priceAt(day) + this.deviation;
  }

  priceTrendPosition(price, day = new Date()) {
    const supportPrice = this.supportPrice(day);
    const difference = price - supportPrice;

    return difference/(this.deviation * 2);
  }

  deviationRatio() {
    return this.deviation / this.priceAt();
  }
}

export default StockTrend.define();
